tools/hive-bigquery/bigquery_table_model.py

```diff
@@ -0,0 +1,180 @@
+            return True
+        return False
+
+    @property
+    def flat_schema(self):
+        if self._flat_schema is None:
+            self._flat_schema = self.flatten_schema()
+        return self._flat_schema
+
+    def flatten_schema(self):
+        """Returns BigQuery table schema in flat structure
+
+        Nested data types in BigQuery schema are represented using nested
+        fields.
+        For example, map column col_name(map<string,int>) is represented as
+        {
+            "fields": [
+            {
+                "mode": "REQUIRED",
+                "name": "key",
+                "type": "STRING"
+            },
+            {
+                "mode": "NULLABLE",
+                "name": "value",
+                "type": "INTEGER"
+            }
+            ],
+            "mode": "REPEATED",
+            "name": "col_name",
+            "type": "RECORD"
+        }
+        To compare the data types in Hive and BigQuery, the schema needs to
+        be flattened and then the data types can be compared.
+
+        For example the above will be flattened as
+        {
+            "col_name"          : "RECORD_REPEATED",
+            "col_name__key"     : "STRING",
+            "col_name__value"   : "INTEGER"
+        }
+        Uses string extraction to flatten the schema"""
+
+        def recursively_flatten(schema, col_name):
+            """Iterates through the nested fields and gets the data types
+
+            Args:
+                schema (List[dict]): schema of the BigQuery fields
+                col_name (str): Flattened column name
+            """
+            for item in schema:
+                name = col_name + item['name']
+                if item['mode'] == 'REPEATED':
+                    col_type = item['type'] + '_' + item['mode']
+                else:
+                    col_type = item['type']
+
+                flat_schema[name] = col_type
+
+                if "RECORD" in col_type:
+                    recursively_flatten(item['fields'], name + '__')
+
+        flat_schema = OrderedDict()
+        recursively_flatten(self.schema, '')
+
+        if self.data_format == "Parquet":
+            for key in flat_schema.keys():
+                find_string = '__bag__array_element'
+                try:
+                    value = flat_schema[key]
+                    if key.endswith(find_string):
+                        flat_schema.pop(key, None)
+                        flat_schema.pop(key[:-len('__array_element')], None)
+
+                        flat_schema[key[:-len(find_string)]] = value
+                        if not value.endswith('_REPEATED'):
+                            flat_schema[key[:-len(find_string)]] += '_REPEATED'
+                    if key.endswith('__map'):
+                        flat_schema.pop(key, None)
+                        flat_schema[key[:-len('__map')]] = value
+
+                except KeyError:
```

Comment:
Explain why you are ignoring this KeyError?

Reply:
@pkattamuri ?

Reply:
I'm iterating through a dictionary and deleting some keys. so, while iterating over a deleted key it will not find that in the dict anymore, so I'm ignoring that. Will see if I can modify that by saving keys of interest in a list